modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/block/BlockService.scala

```diff
@@ -62,35 +58,29 @@ object BlockService {
         } yield ()
 
       private def prepareBalances(transactions: Set[Transaction]): F[Map[Address, Balance]] = {
-        val sources = transactions.groupBy(_.source)
-        val destinations = transactions.groupBy(_.destination)
-        val all = sources.combine(destinations)
-        val addressBalanceChanges = all.map {
-          case (address, txs) =>
-            val change = txs.foldLeft(BigInt(0L))(
-              (change, tx) =>
-                tx match {
-                  case Transaction(`address`, _, amount, fee, _, _) => change - amount.coerce - fee.coerce
-                  case Transaction(_, `address`, amount, _, _, _)   => change + amount.coerce
-                  case _                                            => change
-                }
-            )
+        val sources = transactions.groupBy(_.source).keySet
+        val destinations = transactions.groupBy(_.destination).keySet
 
-            address -> change
-        }
+        val addresses = sources ++ destinations
 
-        addressBalanceChanges.toList.traverse {
-          case (address, balanceChange) =>
-            for {
-              current <- addressStorage.getBalance(address)
-              newBalance <- refineV[NonNegative]
-                .apply(current.coerce + balanceChange)
-                .bimap(BalanceNegative, Balance(_))
-                .liftTo[F]
-            } yield (address, newBalance)
+        def applyTransactions(balance: Balance, address: Address): Either[BalanceOutOfRange, Balance] =
+          transactions.foldLeft(balance.asRight[BalanceOutOfRange]) { (acc, tx) =>
+            tx match {
+              case Transaction(`address`, _, amount, fee, _, _) => acc.flatMap(_.minus(amount)).flatMap(_.minus(fee))
+              case Transaction(_, `address`, amount, _, _, _)   => acc.flatMap(_.plus(amount))
```

Comment:
The problem I see with this adding/subtracting is that it is sensitive to an order in which transactions are folded. Currently that shouldn't be a problem because the validation will not pass transactions if we don't have sufficient balance up front, but in the future if we allow the received transactions to influence our balance during block validation this will no longer be valid. And I guess we may wan't to do this at some point because it seems a valid scenario. So we need to remember that.